File: Leetcode/树/填充每个节点的下一个右侧节点指针.java
class Solution {
    public Node connect(Node root) {
        if (root == null) {
            return root;
        }
        Queue<Node> Q = new LinkedList<Node>(); 
        Q.add(root);
        while (Q.size() > 0) {
            int size = Q.size();
            for(int i = 0; i < size; i++) {
                Node node = Q.poll();
                if (i < size - 1) {
                    node.next = Q.peek();
                }
                if (node.left != null) {
                    Q.add(node.left);
                }
                if (node.right != null) {
                    Q.add(node.right);
                }
            }
        }
        return root;
    }
}

File: Leetcode/树/填充每个节点的下一个右侧节点指针.cpp
class Node{
public:  
    int val;
    Node* left;
    Node* right;
    Node* next;
    Node():val(0),left(nullptr),right(nullptr),next(nullptr) {}
    Node(int _val):val(_val),left(nullptr),right(nullptr),next(nullptr){}
    Node(int _val, Node* _left, Node* _right, Node* _next):val(_val),left(_left),right(_right),next(_next){}
};
class Solution{
public:  
    Node* connect(Node* root){
        if(root==nullptr) return root;
        Node* leftmost = root;
        while(leftmost->left){
            Node* head = leftmost;
            while(head){
                head->left->next = head->right;
                if(head->next){
                    head->right->next = head->next->left;
                }
                head = head->next;
            }
            leftmost = leftmost->right;
        }
        return root;
    }
};